docs/advanced/inspecting-roact-objects.md

```diff
@@ -0,0 +1,71 @@
+In certain situations, such as when building highly reusable and customizable components, props may be composed of Roact objects, such as an element or a component class.
+
+To facilitate safer development for these kinds of situations, Roact exposes the `Roact.typeOf` function to inspect Roact objects and return a value from the `Roact.Type` enumeration.
+
+## Without Object Type Inspection
+
+Suppose we want to write a Header component with a prop for the title child element:
+```lua
+local Header = Component:extend("Header")
+function Header:render()
+	local titleClass = props.titleClass
+	return Roact.createElement("Frame", {
+		-- Props for Frame...
+	}, {
+		Title = Roact.createElement(titleClass, {
+			-- Props for Title...
+		})
+	})
+end
+```
+
+Now suppose we want to validate that titleClass is actually a class using [validateProps](../../api-reference/#validateprops). Unfortunately, the best we can do is query Header to see if it contains characteristics of a Component class:
+```lua
+local Header = Component:extend("Header")
+Header.validateProps = function()
+	local titleClass = props.titleClass
+	if type(titleClass.render) == "function" then
+		return true
+	end
+	return false, tostring(Header) .. " prop titleClass cannot render"
+end
+```
+
+## With Object Type Inspection
+
+With `Roact.typeOf`, we can be certain we have a Component class:
+```lua
+Header.validateProps = function()
+	local titleClass = props.titleClass
+	if Roact.typeOf(titleClass) == Roact.Type.StatefulComponentClass then
```

Comment:
This is something I hadn't thought about before -- do we want people writing checks like this? Since a component can be a `StatefulComponentClass`, any `function`, or any `string`, it feels odd to specifically check for one of these options.

Reply:
It's kind of like bringing ElementKind accessible at this point, but I am not sure if we want that or not.

Reply:
I think you're right. And even `ElementKind` has limited use from a developer's point of view since you have to have already created the element to use it. After talking to @dbrooks-roblox , I think we need to add Roact.isValidElementType in this PR - (basically [this](https://github.com/facebook/react/blob/master/packages/shared/isValidElementType.js)